File: Utilities/Matlab/scripts/mccaffrey_plume.m
% McDermott
% 9-19-2017
% mccaffrey_plume.m

% first call McCaffrey_Correlation.m to make data plots

disp('mccaffrey_plume ...')

McCaffrey_Correlation
McCaffrey_Surface_Temp

close all
clear all

plot_style

Q = [14.4 21.7 33.0 44.9 57.5]; % kW [14.4 21.7 33.0 44.9 57.5]
g = 9.8; % m/s2
rho = 1.18; % kg/m3
cp = 1; % kJ/(kg*K)
T0 = 273.15 + 20; % K
D = 0.3; % m

QS = (Q/(rho*cp*T0*sqrt(g)*D^(5/2)));
DS = (Q/(rho*cp*T0*sqrt(g))).^(2/5); % m

EXP_Dir = '../../../exp/';
OUT_Dir = '../../../out/';
Manuals_Dir = '../../Manuals/';

datadir = [OUT_Dir,'McCaffrey_Plume/'];
plotdir = [Manuals_Dir,'FDS_Validation_Guide/SCRIPT_FIGURES/McCaffrey_Plume/'];

resolution = {'Coarse','Medium','Fine'};

for k=1:length(resolution)

    switch resolution{k}
        case 'Coarse'
            chid = {'McCaffrey_14_kW_11','McCaffrey_22_kW_11','McCaffrey_33_kW_11','McCaffrey_45_kW_11','McCaffrey_57_kW_11'};
        case 'Medium'
            chid = {'McCaffrey_14_kW_21','McCaffrey_22_kW_21','McCaffrey_33_kW_21','McCaffrey_45_kW_21','McCaffrey_57_kW_21'};
        case 'Fine'
            chid = {'McCaffrey_14_kW_45','McCaffrey_22_kW_45','McCaffrey_33_kW_45','McCaffrey_45_kW_45','McCaffrey_57_kW_45'};
    end
    mark = {'ko','k+','k^','ksq','kd'};
    n_chid = length(chid);

    % McCaffrey plume correlations

    zq = logspace(-2,0,100);

    for i=1:length(zq)
        if zq(i)<0.08
            vq(i) = 6.84*zq(i)^0.5;
            Tq(i) = 800*zq(i)^0;
        elseif zq(i)>=0.08 & zq(i)<=0.2
            vq(i) = 1.93*zq(i)^0;
            Tq(i) = 63*zq(i)^(-1);
        elseif zq(i)>0.2
            vq(i) = 1.12*zq(i)^(-1/3);
            Tq(i) = 21.6*zq(i)^(-5/3);
        end
    end

    figure
    set(gca,'Units',Plot_Units)
    set(gca,'Position',[Plot_X Plot_Y Plot_Width Plot_Height])

    hh(n_chid+1)=loglog(zq,vq,'b--','linewidth',2); hold on

    % for McCaffrey 1979 scaling use:
    xmin = 0.01;
    xmax = 1.0;
    ymin = 0.5;
    ymax = 3.0;
    axis([xmin xmax ymin ymax])
    %grid on
    xlabel('{\itz/Q}^{2/5} (m kW^{-2/5})','FontSize',Label_Font_Size,'Interpreter',Font_Interpreter)
    ylabel('{\itV/Q}^{1/5} (m s^{-1} kW^{-1/5})','FontSize',Label_Font_Size,'Interpreter',Font_Interpreter)

    % FDS results velocity

    for i=1:n_chid
        M = importdata([datadir,chid{i},'_line.csv'],',',2);
        z = M.data(:,find(strcmp('Height',M.colheaders)));
        v = M.data(:,find(strcmp('vel',M.colheaders)));

        % McCaffrey 1979 scaling
        zq_fds = z./Q(i)^(2/5);
        vq_fds = v./Q(i)^(1/5);

        hh(i)=loglog(zq_fds,vq_fds,mark{i});
    end

    set(gca,'FontName',Font_Name)
    set(gca,'FontSize',Label_Font_Size)

    leg_key = {'14.4 kW','21.7 kW','33.0 kW','44.9 kW','57.5 kW','({\itz/Q}^{2/5})^\eta'};
    lh = legend(hh,leg_key,'location','southeast');
    set(lh,'Interpreter',Font_Interpreter)
    set(lh,'FontSize',Key_Font_Size)

    text(.0125,2.6,['McCaffrey Centerline Velocity, ',resolution{k}],'FontSize',Title_Font_Size,'FontName',Font_Name)

    text(.04,1.2,'\eta=1/2','FontSize',Label_Font_Size,'FontName',Font_Name)
    text(.10,1.7,'\eta=0','FontSize',Label_Font_Size,'FontName',Font_Name)
    text(.295,1.25,'\eta=-1/3','FontSize',Label_Font_Size,'FontName',Font_Name)

    % add version string if file is available

    Git_Filename = [datadir,'McCaffrey_14_kW_11_git.txt'];
    addverstr(gca,Git_Filename,'loglog')

    % print to pdf

    set(gcf,'Visible',Figure_Visibility);
    set(gcf,'Units',Paper_Units);
    set(gcf,'PaperUnits',Paper_Units);
    set(gcf,'PaperSize',[Paper_Width Paper_Height]);
    set(gcf,'Position',[0 0 Paper_Width Paper_Height]);
    print(gcf,'-dpdf',[plotdir,['McCaffrey_Velocity_Correlation_',resolution{k}]])


    % FDS results temperature rise

    figure
    set(gca,'Units',Plot_Units)
    set(gca,'Position',[Plot_X Plot_Y Plot_Width Plot_Height])

    hh(n_chid+1)=loglog(zq,Tq,'r--','linewidth',2); hold on

    % McCaffrey scaling
    xmin = 0.008;
    xmax = 1.0;
    ymin = 100;
    ymax = 1200;
    axis([xmin xmax ymin ymax])
    %grid on
    xlabel('{\itz/Q}^{2/5} (m kW^{-2/5})','FontSize',Label_Font_Size,'Interpreter',Font_Interpreter)
    ylabel('\Delta{\itT} (\circC)','FontSize',Label_Font_Size,'Interpreter',Font_Interpreter)

    for i=1:n_chid
        M = importdata([datadir,chid{i},'_line.csv'],',',2);
        z = M.data(:,find(strcmp('Height',M.colheaders)));
        T = M.data(:,find(strcmp('tmp',M.colheaders))) + 273.15;
        zq_fds = z./Q(i)^(2/5);
        hh(i)=loglog(zq_fds,T-T0,mark{i});
    end

    set(gca,'FontName',Font_Name)
    set(gca,'FontSize',Label_Font_Size)

    leg_key = {'14.4 kW','21.7 kW','33.0 kW','44.9 kW','57.5 kW','({\itz/Q}^{2/5})^\eta'};
    lh = legend(hh,leg_key,'location','southwest');
    set(lh,'Interpreter',Font_Interpreter)
    set(lh,'FontSize',Key_Font_Size)

    text(.01,1000,['McCaffrey Centerline Temperature, ',resolution{k}],'FontSize',Title_Font_Size,'FontName',Font_Name)

    text(.02,650,'\eta=0','FontSize',Label_Font_Size,'FontName',Font_Name)
    text(.08,425,'\eta=-1','FontSize',Label_Font_Size,'FontName',Font_Name)
    text(.18,125,'\eta=-5/3','FontSize',Label_Font_Size,'FontName',Font_Name)

    % add version string if file is available

    Git_Filename = [datadir,'McCaffrey_14_kW_11_git.txt'];
    addverstr(gca,Git_Filename,'loglog')

    % print to pdf

    set(gcf,'Visible',Figure_Visibility);
    set(gcf,'Units',Paper_Units);
    set(gcf,'PaperUnits',Paper_Units);
    set(gcf,'PaperSize',[Paper_Width Paper_Height]);
    set(gcf,'Position',[0 0 Paper_Width Paper_Height]);
    print(gcf,'-dpdf',[plotdir,['McCaffrey_Temperature_Correlation_',resolution{k}]])

end

%%% CURRENT INPUT FILES DO NOT HAVE DEVC FOR PROFILES %%%
return


% Velocity profile in the plume

figure
set(gca,'Units',Plot_Units)
set(gca,'Position',[Plot_X Plot_Y Plot_Width Plot_Height])

z = 1;
chid = {'McCaffrey_22_kW_11','McCaffrey_22_kW_21','McCaffrey_22_kW_45'};
mark = {'kd','k^','ksq'};
for i=1:length(chid)
    M = importdata([datadir,chid{i},'_line.csv'],',',2);
    x = M.data(:,find(strcmp('Radius',M.colheaders)));
    V = M.data(:,find(strcmp('wx_p',M.colheaders)));
    V0 = V(find(x==0));
    hh(i)=plot(x/z,V/V0,mark{i}); hold on
end

z = 1.25;
chid = {'McCaffrey_45_kW_11','McCaffrey_45_kW_21','McCaffrey_45_kW_45'};
mark = {'kd','k^','ksq'};
for i=1:length(chid)
    M = importdata([datadir,chid{i},'_line.csv'],',',2);
    x = M.data(:,find(strcmp('Radius',M.colheaders)));
    V = M.data(:,find(strcmp('wx_p',M.colheaders)));
    V0 = V(find(x==0));
    hh(length(chid)+i)=plot(x/z,V/V0,mark{i},'MarkerFaceColor','k'); hold on
end

x = linspace(0,.3,100);
a1 = 0.16; % see Fig. 6, McCaffrey (1979)
a2 = 0.14;
a3 = 0.18;
v1 = exp(-(5/6/a1*x).^2);
v2 = exp(-(5/6/a2*x).^2);
v3 = exp(-(5/6/a3*x).^2);
hh(2*length(chid)+1)=plot(x,v1,'k-');
hh(2*length(chid)+2)=plot(x,v2,'k--');
hh(2*length(chid)+3)=plot(x,v3,'k-.');

set(gca,'FontName',Font_Name)
set(gca,'FontSize',Label_Font_Size)

axis([0 .35 0 1.2])
xlabel('$x/z$','FontSize',Label_Font_Size,'Interpreter',Font_Interpreter)
ylabel('$V/V_0$','FontSize',Label_Font_Size,'Interpreter',Font_Interpreter)

leg_key = {'22 kW 11','22 kW 21','22 kW 45',...
           '45 kW 11','45 kW 21','45 kW 45',...,
           '$\alpha=0.14$','$\alpha=0.16$','$\alpha=0.18$'};
lh=legend(hh,leg_key,'location','northeast');
set(lh,'Interpreter',Font_Interpreter)
set(lh,'FontSize',Key_Font_Size)

xt = .05;
yt = 1.1;
text(xt,yt,'Plume','FontSize',Title_Font_Size,'Interpreter',Font_Interpreter,'FontName',Font_Name)

% add version string if file is available

Git_Filename = [datadir,'McCaffrey_14_kW_11_git.txt'];
addverstr(gca,Git_Filename,'linear')

% print to pdf

set(gcf,'Visible',Figure_Visibility);
set(gcf,'Units',Paper_Units);
set(gcf,'PaperUnits',Paper_Units);
set(gcf,'PaperSize',[Paper_Width Paper_Height]);
set(gcf,'Position',[0 0 Paper_Width Paper_Height]);
print(gcf,'-dpdf',[plotdir,'McCaffrey_Velocity_Profile_Plume'])


% Velocity profile in the intermittent region

figure
set(gca,'Units',Plot_Units)
set(gca,'Position',[Plot_X Plot_Y Plot_Width Plot_Height])

clear hh

z = 0.5;
x_1oe = 0.46*(z/Q(2)^(2/5)) + 0.013*sqrt(Q(2));
chid = {'McCaffrey_22_kW_11','McCaffrey_22_kW_21','McCaffrey_22_kW_45'};
mark = {'kd','k^','ksq'};
for i=1:length(chid)
    M = importdata([datadir,chid{i},'_line.csv'],',',2);
    x = M.data(:,find(strcmp('Radius',M.colheaders)));
    V = M.data(:,find(strcmp('wx_i',M.colheaders)));
    V0 = V(find(x==0));
    hh(i)=plot(x/x_1oe,V/V0,mark{i}); hold on
end

z = 0.75;
x_1oe = 0.46*(z/Q(4)^(2/5)) + 0.013*sqrt(Q(4));
chid = {'McCaffrey_45_kW_11','McCaffrey_45_kW_21','McCaffrey_45_kW_45'};
mark = {'kd','k^','ksq'};
for i=1:length(chid)
    M = importdata([datadir,chid{i},'_line.csv'],',',2);
    x = M.data(:,find(strcmp('Radius',M.colheaders)));
    V = M.data(:,find(strcmp('wx_i',M.colheaders)));
    V0 = V(find(x==0));
    hh(length(chid)+i)=plot(x/x_1oe,V/V0,mark{i},'MarkerFaceColor','k'); hold on
end

x = linspace(0,2,100);
v1 = exp(-x.^1.5);
hh(2*length(chid)+1)=plot(x,v1,'k-');

set(gca,'FontName',Font_Name)
set(gca,'FontSize',Label_Font_Size)

axis([0 3 0 1.2])
xlabel('$x/x_{1/e}$','FontSize',Label_Font_Size,'Interpreter',Font_Interpreter)
ylabel('$V/V_0$','FontSize',Label_Font_Size,'Interpreter',Font_Interpreter)

leg_key = {'22 kW 11','22 kW 21','22 kW 45',...
           '45 kW 11','45 kW 21','45 kW 45',...
           '$n=3/2$'};
lh=legend(hh,leg_key,'location','northeast');
set(lh,'Interpreter',Font_Interpreter)
set(lh,'FontSize',Key_Font_Size)

xt = .5;
yt = 1.1;
text(xt,yt,'Intermittent','FontSize',Title_Font_Size,'Interpreter',Font_Interpreter,'FontName',Font_Name)

% add version string if file is available

Git_Filename = [datadir,'McCaffrey_14_kW_11_git.txt'];
addverstr(gca,Git_Filename,'linear')

% print to pdf

set(gcf,'Visible',Figure_Visibility);
set(gcf,'Units',Paper_Units);
set(gcf,'PaperUnits',Paper_Units);
set(gcf,'PaperSize',[Paper_Width Paper_Height]);
set(gcf,'Position',[0 0 Paper_Width Paper_Height]);
print(gcf,'-dpdf',[plotdir,'McCaffrey_Velocity_Profile_Intermittent'])


% Velocity profile in the flame region

figure
set(gca,'Units',Plot_Units)
set(gca,'Position',[Plot_X Plot_Y Plot_Width Plot_Height])

clear hh

z = 0.125;
x_1oe = 0.46*(z/Q(2)^(2/5)) + 0.013*sqrt(Q(2));
chid = {'McCaffrey_22_kW_11','McCaffrey_22_kW_21','McCaffrey_22_kW_45'};
mark = {'kd','k^','ksq'};
for i=1:length(chid)
    M = importdata([datadir,chid{i},'_line.csv'],',',2);
    x = M.data(:,find(strcmp('Radius',M.colheaders)));
    V = M.data(:,find(strcmp('wx_f',M.colheaders)));
    V0 = V(find(x==0));
    hh(i)=plot(x/x_1oe,V/V0,mark{i}); hold on
end

z = 0.25;
x_1oe = 0.46*(z/Q(4)^(2/5)) + 0.013*sqrt(Q(4));
chid = {'McCaffrey_45_kW_11','McCaffrey_45_kW_21','McCaffrey_45_kW_45'};
mark = {'kd','k^','ksq'};
for i=1:length(chid)
    M = importdata([datadir,chid{i},'_line.csv'],',',2);
    x = M.data(:,find(strcmp('Radius',M.colheaders)));
    V = M.data(:,find(strcmp('wx_f',M.colheaders)));
    V0 = V(find(x==0));
    hh(length(chid)+i)=plot(x/x_1oe,V/V0,mark{i},'MarkerFaceColor','k'); hold on
end

x = linspace(0,2,100);
v1 = exp(-x.^1.5);
v2 = exp(-x.^2);
hh(2*length(chid)+1)=plot(x,v1,'k-');
hh(2*length(chid)+2)=plot(x,v2,'k--');

set(gca,'FontName',Font_Name)
set(gca,'FontSize',Label_Font_Size)

axis([0 3 0 1.2])
xlabel('$x/x_{1/e}$','FontSize',Label_Font_Size,'Interpreter',Font_Interpreter)
ylabel('$V/V_0$','FontSize',Label_Font_Size,'Interpreter',Font_Interpreter)

leg_key = {'22 kW 11','22 kW 21','22 kW 45',...
           '45 kW 11','45 kW 21','45 kW 45',...
           '$n=3/2$','$n=2$'};
lh=legend(hh,leg_key,'location','northeast');
set(lh,'Interpreter',Font_Interpreter)
set(lh,'FontSize',Key_Font_Size)

xt = .5;
yt = 1.1;
text(xt,yt,'Flame','FontSize',Title_Font_Size,'Interpreter',Font_Interpreter,'FontName',Font_Name)

% add version string if file is available

Git_Filename = [datadir,'McCaffrey_14_kW_11_git.txt'];
addverstr(gca,Git_Filename,'linear')

% print to pdf

set(gcf,'Visible',Figure_Visibility);
set(gcf,'Units',Paper_Units);
set(gcf,'PaperUnits',Paper_Units);
set(gcf,'PaperSize',[Paper_Width Paper_Height]);
set(gcf,'Position',[0 0 Paper_Width Paper_Height]);
print(gcf,'-dpdf',[plotdir,'McCaffrey_Velocity_Profile_Flame'])

File: Utilities/Python/scripts/McCaffrey_Plume.py

import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# include FDS plot styles, etc.
import fdsplotlib

# Get plot style parameters
plot_style = fdsplotlib.get_plot_style('fds')

expdir = '../../../exp/Submodules/macfp-db/Gaseous_Pool_Fires/McCaffrey_Flames/Computational_Results/2017/Data/'
outdir = '../../../out/'
Manuals_Dir = '../../Manuals/'
datadir = os.path.join(outdir, 'McCaffrey_Plume')
pltdir = os.path.join(Manuals_Dir, 'FDS_Validation_Guide', 'SCRIPT_FIGURES', 'McCaffrey_Plume')
git_file =  os.path.join(datadir, f'McCaffrey_14_kW_11_git.txt')
version_string = fdsplotlib.get_version_string(git_file)

Q = np.array([14.4, 21.7, 33.0, 44.9, 57.5])  # kW
chid = ['McCaffrey_14kW', 'McCaffrey_22kW', 'McCaffrey_33kW', 'McCaffrey_45kW', 'McCaffrey_57kW']
mark = ['ko', 'k+', 'k^', 'ks', 'kd']

# Velocity and temperature correlations

zq = np.logspace(-2, 0, 100)
vq = np.zeros_like(zq)
Tq = np.zeros_like(zq)
for i, z in enumerate(zq):
    if z < 0.08:
        vq[i] = 6.84 * z**0.5
        Tq[i] = 800.0
    elif 0.08 <= z <= 0.2:
        vq[i] = 1.93
        Tq[i] = 63.0 * z**(-1)
    else:
        vq[i] = 1.12 * z**(-1.0/3.0)
        Tq[i] = 21.6 * z**(-5.0/3.0)

# Plot velocity correlation and measurement points

fig1 = fdsplotlib.plot_to_fig(x_data=zq, y_data=vq, data_label='$(z/Q^{2/5})^\\eta$', 
                              marker_style='b--',
                              x_min=0.01, x_max=1, y_min=0.5, y_max=3,
                              x_label='$z/Q^{2/5}$ $(\\hbox{m kW}^{-2/5})$',
                              y_label='${V/Q}^{1/5}$ (m s$^{-1}$ kW$^{-1/5}$)',
                              legend_location='lower right',
                              plot_type='loglog')

for i, c in enumerate(chid):
    fname = os.path.join(expdir, f'{c}_V.csv')
    if os.path.exists(fname):
        M = pd.read_csv(fname, comment=None)
        fdsplotlib.plot_to_fig(x_data=M.iloc[:,0].values,y_data=M.iloc[:, 1].values, 
                               figure_handle=fig1,
                               marker_style=mark[i], data_label=f'{Q[i]:.1f} kW')

ax = plt.gca()
ax.text(0.012,1.2, '$\\eta=1/2$' , fontsize=plot_style['Label_Font_Size'])
ax.text(0.100,2.3, '$\\eta=0$'   , fontsize=plot_style['Label_Font_Size'])
ax.text(0.400,1.8, '$\\eta=-1/3$', fontsize=plot_style['Label_Font_Size'])

fig1.savefig(os.path.join(pltdir, 'McCaffrey_Velocity_Correlation.pdf'), format='pdf')
plt.close(fig1)

# Plot temperature correlation and measurement points

fig2 = fdsplotlib.plot_to_fig(x_data=zq, y_data=Tq, data_label='$(z/Q^{2/5})^\\eta$',
                              marker_style='r--',
                              x_min=0.008, x_max=1, y_min=100, y_max=1200,
                              x_label='$z/Q^{2/5}$ $(\\hbox{m kW}^{-2/5})$',
                              y_label='$\\Delta T \\; (^\\circ$C)',
                              legend_location='lower left',
                              plot_type='loglog')

for i, c in enumerate(chid):
    fname = os.path.join(expdir, f'{c}_T.csv')
    if os.path.exists(fname):
        M = pd.read_csv(fname, comment=None)
        fdsplotlib.plot_to_fig(x_data=M.iloc[:,0].values,y_data=M.iloc[:, 1].values,
                               figure_handle=fig2,
                               marker_style=mark[i], data_label=f'{Q[i]:.1f} kW')

ax = plt.gca()
ax.text(0.012,500, '$\\eta=0$'   , fontsize=plot_style['Label_Font_Size'])
ax.text(0.060,400, '$\\eta=-1$'  , fontsize=plot_style['Label_Font_Size'])
ax.text(0.100,150, '$\\eta=-5/3$', fontsize=plot_style['Label_Font_Size'])

fig2.savefig(os.path.join(pltdir, 'McCaffrey_Temperature_Correlation.pdf'), format='pdf')
plt.close(fig2)

# Extrapolate temperature correlation to burner surface

os.makedirs(pltdir, exist_ok=True)
for i, c in enumerate(chid):
    fname = os.path.join(expdir, f'{c}_T.csv')
    if not os.path.exists(fname):
        continue
    M = pd.read_csv(fname, comment=None)
    ZS = M.iloc[:, 0].values
    TC = M.iloc[:, 1].values

    fig = fdsplotlib.plot_to_fig(x_data=zq, y_data=Tq, data_label='$(z/Q^{2/5})^\\eta$',
                                 marker_style='r--',
                                 x_min=0.008, x_max=1, y_min=100, y_max=1200,
                                 x_label='$z/Q^{2/5}$ $(\\hbox{m kW}^{-2/5})$',
                                 y_label='$\\Delta T \\; (^\\circ$C)',
                                 legend_location='lower left',
                                 plot_title=f'McCaffrey Centerline Temperature Data, {Q[i]:.1f} kW',
                                 plot_type='loglog')

    fdsplotlib.plot_to_fig(x_data=ZS,y_data=TC,
                           figure_handle=fig,
                           marker_style=mark[i], data_label=f'{Q[i]:.1f} kW')

    # identify near-surface (ZS < 0.05)
    isurf = np.where(ZS < 0.05)[0]
    if len(isurf) >= 2:
        fdsplotlib.plot_to_fig(x_data=ZS[isurf], y_data=TC[isurf], marker_style='bo', figure_handle=fig)
        # least squares fit (linear in this small z region)
        A = np.column_stack((ZS[isurf], np.ones_like(ZS[isurf])))
        b = TC[isurf]
        x, *_ = np.linalg.lstsq(A, b, rcond=None)
        T_surf = x[1]
        z1 = 0.008; z2 = 0.05
        T1 = x[0]*z1 + x[1]
        T2 = x[0]*z2 + x[1]
        fdsplotlib.plot_to_fig(x_data=[z1,z2], y_data=[T1,T2], marker_style='g-', figure_handle=fig, data_label='Least Squares Fit')
        ax = plt.gca()
        ax.text(0.009, 400., '$T_{\\rm surf}$='+f'{T_surf:3.0f} $^\\circ$C', fontsize=plot_style['Label_Font_Size'])
    else:
        T_surf = np.nan

    fig.savefig(os.path.join(pltdir, f'{c}_Surface_Temp.pdf'), format='pdf')
    plt.close(fig)

# Plot FDS predictions

T0 = 273.15 + 20  # K

resolution = ['Crude', 'Coarse', 'Medium', 'Fine']
for res in resolution:
    if res == 'Crude':
        chid = ['McCaffrey_14_kW_5','McCaffrey_22_kW_5','McCaffrey_33_kW_5','McCaffrey_45_kW_5','McCaffrey_57_kW_5']
    elif res == 'Coarse':
        chid = ['McCaffrey_14_kW_11','McCaffrey_22_kW_11','McCaffrey_33_kW_11','McCaffrey_45_kW_11','McCaffrey_57_kW_11']
    elif res == 'Medium':
        chid = ['McCaffrey_14_kW_21','McCaffrey_22_kW_21','McCaffrey_33_kW_21','McCaffrey_45_kW_21','McCaffrey_57_kW_21']
    elif res == 'Fine':
        chid = ['McCaffrey_14_kW_45','McCaffrey_22_kW_45','McCaffrey_33_kW_45','McCaffrey_45_kW_45','McCaffrey_57_kW_45']

    # Velocity correlation figures (one for each resolution)

    fig_v = fdsplotlib.plot_to_fig(x_data=zq, y_data=vq, data_label='$(z/Q^{2/5})^\\eta$',
                                   marker_style='b--',
                                   x_min=0.01, x_max=1, y_min=0.5, y_max=3,
                                   revision_label=version_string,
                                   x_label='$z/Q^{2/5}$ $(\\hbox{m kW}^{-2/5})$',
                                   y_label='$V/Q^{1/5}$ $(\\hbox{m s}^{-1} \\; \\hbox{kW}^{-1/5})$',
                                   plot_title=f'McCaffrey Centerline Velocity, {res}',
                                   legend_location='lower right',
                                   plot_type='loglog')

    for i, c in enumerate(chid):
        fname = os.path.join(datadir, f'{c}_line.csv')
        if os.path.exists(fname):
            M = pd.read_csv(fname, header=1)
            # find columns by name if present; else use first/second
            if 'Height' in M.columns and 'vel' in M.columns:
                z = M['Height'].values
                v = M['vel'].values
            else:
                z = M.iloc[:, 0].values
                v = M.iloc[:, 1].values
            zq_fds = z / (Q[i]**(2.0/5.0))
            vq_fds = v / (Q[i]**(1.0/5.0))
            fdsplotlib.plot_to_fig(x_data=zq_fds, y_data=vq_fds, 
                                          data_label=f'{Q[i]} kW', marker_style=mark[i], figure_handle=fig_v)
    fig_v.savefig(os.path.join(pltdir, f'McCaffrey_Velocity_Correlation_{res}.pdf'))
    plt.close(fig_v)

    # Temperature correlation figures (one for each resolution)

    fig_t = fdsplotlib.plot_to_fig(x_data=zq, y_data=Tq, data_label='$(z/Q^{2/5})^\\eta$',
                                   marker_style='r--',
                                   x_min=0.008, x_max=1, y_min=100, y_max=1200,
                                   revision_label=version_string,
                                   x_label='$z/Q^{2/5}$ $(\\hbox{m kW}^{-2/5})$',
                                   y_label='$\\Delta{T}$ ($^\\circ$C)',
                                   plot_title=f'McCaffrey Centerline Temperature, {res}',
                                   legend_location='lower right',
                                   plot_type='loglog')

    for i, c in enumerate(chid):
        fname = os.path.join(datadir, f'{c}_line.csv')
        if os.path.exists(fname):
            M = pd.read_csv(fname, header=1)
            if 'Height' in M.columns and 'tmp' in M.columns:
                z = M['Height'].values
                T = M['tmp'].values + 273.15
            else:
                z = M.iloc[:, 0].values
                T = M.iloc[:, 2].values + 273.15 if M.shape[1] > 2 else M.iloc[:, 1].values + 273.15
            zq_fds = z / (Q[i]**(2.0/5.0))
            fdsplotlib.plot_to_fig(x_data=zq_fds, y_data=T-T0, 
                                          data_label=f'{Q[i]} kW', marker_style=mark[i], figure_handle=fig_t)
    fig_t.savefig(os.path.join(pltdir, f'McCaffrey_Temperature_Correlation_{res}.pdf'))
    plt.close(fig_t)

